File: GAMECARO/src/GameRenderer.cpp
// ...existing code...
#include "Game/GameRenderer.h"

using namespace Client::Game;

GameRenderer::GameRenderer(Panel^ canvas, Board^ board)
{
    _canvas = canvas;
    _board = board;
    WinningCells = nullptr;
    _canvas->Paint += gcnew PaintEventHandler(this, &GameRenderer::OnPaint);
}

Nullable<Point> GameRenderer::PointToCell(Point pt)
{
    // bảo vệ chia cho 0 và trả Point với X=col, Y=row (consistent)
    if (_board == nullptr || _board->Cols <= 0 || _board->Rows <= 0 || _canvas == nullptr) 
        return Nullable<Point>();

    int cols = _board->Cols;
    int rows = _board->Rows;
    int cellW = _canvas->Width / cols;
    int cellH = _canvas->Height / rows;
    if (cellW <= 0 || cellH <= 0) return Nullable<Point>();

    int col = pt.X / cellW;
    int row = pt.Y / cellH;
    if (row < 0 || row >= rows || col < 0 || col >= cols) return Nullable<Point>();

    // X = column, Y = row
    return Point(col, row);
}

void GameRenderer::Refresh()
{
    if (_canvas != nullptr) _canvas->Invalidate();
}

void GameRenderer::OnPaint(Object^ sender, PaintEventArgs^ e)
{
    if (_canvas == nullptr || _board == nullptr) return;

    Graphics^ g = e->Graphics;
    int w = _canvas->Width;
    int h = _canvas->Height;
    int rows = _board->Rows;
    int cols = _board->Cols;
    if (rows <= 0 || cols <= 0) return;

    int cellW = w / cols;
    int cellH = h / rows;
    if (cellW <= 0 || cellH <= 0) return;

    Pen^ gridPen = gcnew Pen(Color::Black);
    // draw horizontal and vertical lines
    for (int r = 0; r <= rows; r++)
        g->DrawLine(gridPen, 0, r * cellH, w, r * cellH);
    for (int c = 0; c <= cols; c++)
        g->DrawLine(gridPen, c * cellW, 0, c * cellW, h);

    // draw marks
    for (int r = 0; r < rows; r++)
    {
        for (int c = 0; c < cols; c++)
        {
            String^ mark = _board->Cells[r, c];
            if (mark == nullptr) continue;

            Rectangle rect(c * cellW, r * cellH, cellW, cellH);

            if (mark->Equals("X"))
            {
                Pen^ xPen = gcnew Pen(Color::Red, 2);
                // both diagonals
                g->DrawLine(xPen, rect.Left, rect.Top, rect.Right, rect.Bottom);
                g->DrawLine(xPen, rect.Left, rect.Bottom, rect.Right, rect.Top);
            }
            else // assume "O" or other -> draw circle/ellipse with padding
            {
                Pen^ oPen = gcnew Pen(Color::Blue, 2);
                int pad = Math::Max(2, Math::Min(cellW, cellH) / 8);
                Rectangle inset(rect.X + pad, rect.Y + pad, Math::Max(0, rect.Width - 2 * pad), Math::Max(0, rect.Height - 2 * pad));
                if (inset.Width > 0 && inset.Height > 0)
                    g->DrawEllipse(oPen, inset);
            }
        }
    }

    // draw winning line if any (WinningCells: X=col, Y=row)
    if (WinningCells != nullptr && WinningCells->Count >= 2)
    {
        Pen^ winPen = gcnew Pen(Color::Red, 3);
        Point start = WinningCells[0];
        Point end = WinningCells[WinningCells->Count - 1];
        int sx = start.X * cellW + cellW / 2;
        int sy = start.Y * cellH + cellH / 2;
        int ex = end.X * cellW + cellW / 2;
        int ey = end.Y * cellH + cellH / 2;
        g->DrawLine(winPen, sx, sy, ex, ey);
    }
}
// ...existing code...

File: GAMECARO/Game/GameRenderer.cs
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;

namespace Client.Game
{
    public class GameRenderer
    {
        private readonly Panel _canvas;
        private readonly Board _board;
        private const int CellSize = 35;

        // 🔴 Danh sách ô thắng (Point.X = col, Point.Y = row)
        public List<Point>? WinningCells { get; set; }

        public GameRenderer(Panel canvas, Board board)
        {
            _canvas = canvas;
            _board = board;
            _canvas.Paint += Canvas_Paint;
        }

        private void Canvas_Paint(object? sender, PaintEventArgs e)
        {
            var g = e.Graphics;
            g.Clear(Color.White);

            Pen gridPen = Pens.Gray;

            // ===== Vẽ lưới =====
            for (int i = 0; i <= Board.Size; i++)
            {
                g.DrawLine(gridPen, i * CellSize, 0, i * CellSize, Board.Size * CellSize);
                g.DrawLine(gridPen, 0, i * CellSize, Board.Size * CellSize, i * CellSize);
            }

            using var font = new Font("Arial", 20, FontStyle.Bold);
            using var brushX = new SolidBrush(Color.Blue);
            using var brushO = new SolidBrush(Color.Red);

            // ===== Vẽ X / O =====
            for (int r = 0; r < Board.Size; r++)
            {
                for (int c = 0; c < Board.Size; c++)
                {
                    int cell = _board.GetCell(r, c);
                    if (cell == 0) continue;

                    string mark = (cell == 1) ? "X" : "O";
                    var x = c * CellSize + 6;
                    var y = r * CellSize + 4;
                    g.DrawString(mark, font, mark == "X" ? brushX : brushO, x, y);
                }
            }

            // ===== Vẽ gạch đỏ khi có 5 ô thắng =====
            if (WinningCells != null && WinningCells.Count >= 2)
            {
                // Mặc định: Point.X = col, Point.Y = row
                Point first = WinningCells[0];
                Point last = WinningCells[WinningCells.Count - 1];

                Point p1 = GetCellCenter(first.Y, first.X); // (row, col)
                Point p2 = GetCellCenter(last.Y, last.X);

                using (var pen = new Pen(Color.Red, 4))
                {
                    pen.StartCap = System.Drawing.Drawing2D.LineCap.Round;
                    pen.EndCap = System.Drawing.Drawing2D.LineCap.Round;
                    g.DrawLine(pen, p1, p2);
                }
            }
        }

        // Tính tâm ô (row, col) để kẻ đường qua giữa
        private Point GetCellCenter(int row, int col)
        {
            int x = col * CellSize + CellSize / 2;
            int y = row * CellSize + CellSize / 2;
            return new Point(x, y);
        }

        // Chuyển tọa độ chuột -> (row, col)
        public (int row, int col)? PointToCell(Point p)
        {
            int col = p.X / CellSize;
            int row = p.Y / CellSize;

            if (row >= 0 && row < Board.Size && col >= 0 && col < Board.Size)
                return (row, col);

            return null;
        }

        public void Refresh() => _canvas.Invalidate();
    }
}
